src/Common/ZooKeeper/ZooKeeperImpl.cpp

```diff
@@ -446,8 +446,16 @@ void ZooKeeper::connect(
                 }
                 connected = true;
 
-                if (connected_callback.has_value())
-                    (*connected_callback)(i, node);
+                /// We can't user 'i' as idx because the nodes have already been shuffled.
+                for (size_t idx = 0; idx < args.hosts.size(); ++idx)
+                {
+                    /// The domain names of two nodes in a cluster cannot point to the same IP address.
+                    Poco::Net::SocketAddress addr(args.hosts[idx]);
+                    if (addr.toString() != node.address.toString())
+                        continue;
```

Comment:
AFAIU, this is wrong, see https://github.com/ClickHouse/ClickHouse/pull/50424#discussion_r1272481642

Reply:
> AFAIU, this is wrong, see [#50424 (comment)](https://github.com/ClickHouse/ClickHouse/pull/50424#discussion_r1272481642)

No, I have updated the code according to [your previous comment](https://github.com/ClickHouse/ClickHouse/pull/50702#discussion_r1222911033). Now, both are addresses.

Reply:
Sorry, I did not notice that. However, I still don't like resolving hostnames and comparing addresses. We do know the list of hosts, and we do know which one we selected to connect to. Why do we need to iterate through the list and resolve hostnames? We only need to save the correct index when selecting a node to connect to. Let's just save the index before shuffling (for example, we can add `size_t original_index` field to `Node` struct, fill it before shufling)

Your approach raises a few questions:
 - Is it possible that we will not find a matching address? Should we throw a LOGICAL_ERROR if we don't find it?
 - https://github.com/ClickHouse/ClickHouse/pull/50702#discussion_r1358308611 
 - https://github.com/ClickHouse/ClickHouse/pull/50702#discussion_r1358318139

Reply:
Done.